### PR TITLE
Let MutableMapping.pop take a different default type.

### LINKED ISSUE #278: Mapping.{get,pop} can return default type
See #223. That PR was reverted because it caused some type errors that it shouldn't, but it's the right thing to do and we should figure out how to make it work.

@gvanrossum , what did the errors you saw (after python/mypy#1595 was fixed) look like?

### PR DESCRIPTION
**Pull request by euresti**
This is particularly useful for the idiom `d.pop(k, None)` to remove an item if it exists.

Fixes #278

### CODE REVIEW

File: stdlib/3/builtins.pyi [hunk 1]
```diff
@@ -610,7 +610,6 @@ class dict(MutableMapping[_KT, _VT], Generic[_KT, _VT]):
     def __init__(self, iterable: Iterable[Tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
     def clear(self) -> None: ...
     def copy(self) -> Dict[_KT, _VT]: ...
-    def pop(self, k: _KT, default: _VT = None) -> _VT: ...
```
Comment (ambv) [inline stdlib/3/builtins.pyi]: Seems a bit arbitrary to remove pop but to leave `clear()`, etc. We should prune more methods in a follow-up diff.

**Comment by ambv** [comment]
This is implemented paralleling what `get()` does.

**Comment by JelleZijlstra** [comment]
This makes `.pop()` return a Union, shouldn't we use an `overload` instead to parallel `.get()`? With a Union users will have to `isinstance` check every usage.

**Comment by euresti** [comment]
I think `Union[_T, _T]` automatically resolves to `_T`  At least that's what reveal_type show.

```
from typing import TypeVar, Union

_T = TypeVar('_T')

def pop(var: Union[int, _T]) -> Union[int, _T]:
    return var

reveal_type(pop(7))  # Revealed type is 'builtins.int'
reveal_type(pop('foo')) # Revealed type is 'Union[builtins.int, builtins.str*]'
reveal_type(pop(None)) # Revealed type is 'Union[builtins.int, builtins.None]'
```

**Comment by euresti** [comment]
Oh wait. pop has a default value.  Let me look.

**Pull request merged**